ffi/athcon/bindings/rust/athcon-client/src/lib.rs

```diff
@@ -95,10 +90,8 @@ impl AthconVm {
 }
 
 pub fn create() -> AthconVm {
-  unsafe {
-    AthconVm {
-      handle: athcon_create_athenavmwrapper(),
-      host_interface: Box::into_raw(Box::new(host::get_athcon_host_interface())),
-    }
+  AthconVm {
+    handle: athena_vmlib::athcon_create_athenavmwrapper(),
```

Comment:
This is cleaner, but I'm not 100% sure that it achieves the original purpose, which is to test the FFI functions end to end over FFI. This isn't actually going over FFI, right? I.e., it's one monolithic program.

Reply:
IMHO, it works the same as before. It's just importing the extern "C" function explicitly instead of hacking around with externs. The following extern declaration:

```rust
extern "C" {
  fn athcon_create_athenavmwrapper() -> *mut ffi::athcon_vm;
}
```

can **only** work when the VM library is linked with the program that wants to call `athcon_create_athenavmwrapper` or `AthconVm::new()` (which calls the former indirectly).

This test will not work because of linker issues (the extern function is declared, but not defined (its code cannot be found):

```rust
#[test]
fn create_vm() {
	AthconVm::new();
}
```

```sh
error: linking with `cc` failed: exit status: 1
  |
  = note: rust-lld: error: undefined symbol: athcon_create_athenavmwrapper
          >>> referenced by lib.rs:23 (ffi/athcon/bindings/rust/athcon-client/src/lib.rs:23)
          >>>               /home/bartosz/workspace/athena/target/debug/deps/athcon_client-80f6a5be35bcb197.a8o110s31e0sc2ia9h8o2pqub.rcgu.o:(athcon_client::AthconVm::new::h9f1953dc58903478)
          collect2: error: ld returned 1 exit status
```

So, 
>I.e., it's one monolithic program

Yes, both ways require the vmlib to be linked during compile time into "one program".


I think the question is how `athcon-client` is supposed to be used. If it's meant to be used with the `vmlib` only, I think this approach is better. However, if it's meant to be used with other implementations of the VM (coming in form of a shared library), then I think it should support opening the libvm.so and calling the `athcon_create_athenavmwrapper` from it (just like the Go bindings do).

Btw, what is the usecase of having Rust bindings in the first place? I feel that, as we have only 1 implementation of the AthenaVM written in Rust, the "user code" should just use the VM code directly, skipping the FFI layer. (using FFI is like Rust -> C -> Rust).

Reply:
> I think the question is how athcon-client is supposed to be used. If it's meant to be used with the vmlib only, I think this approach is better. However, if it's meant to be used with other implementations of the VM (coming in form of a shared library), then I think it should support opening the libvm.so and calling the athcon_create_athenavmwrapper from it (just like the Go bindings do).

For now it's only used in the `ffitest` tests. It's also there as a sort of reference code, if someone wants to use Athena with a Rust-based host (e.g., a future rust-spacemesh client).

I have no issue if we want to rewrite this to work like the Go bindings, but it's definitely not high priority.

> Btw, what is the usecase of having Rust bindings in the first place? I feel that, as we have only 1 implementation of the AthenaVM written in Rust, the "user code" should just use the VM code directly, skipping the FFI layer. (using FFI is like Rust -> C -> Rust).

The Rust bindings are strictly for tests (and, as mentioned, reference code). IMHO it's just nice to have a buttoned-up e2e test of the FFI paths self-contained in this repository, so we have some confidence that changes here won't cause downstream issues for FFI clients. Of course we could write something simple in C or even Go to do this, if you think that's better.

Reply:
I double-checked: both ways of calling `athcon_create_athenavmwrapper` **don't** actually go over FFI. They are normal Rust function calls. To enforce going through the FFI, the test would need to open a VM shared library and call the function from it.

However, I think it's not important here. These tests aim to test the exported C interface of the vmlib, not the Rust FFI calling, and they achieve this goal.